docs/source/developer_guides/model_merging.md

```diff
@@ -0,0 +1,178 @@
+<!--Copyright 2024 The HuggingFace Team. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+the License. You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# Model merging
+
+Training a model for each task can be costly, take up storage space, and the models aren't able to learn new information to improve their performance. Multitask learning can overcome some of these limitations by training a model to learn several tasks, but this is expensive to train and designing a dataset for it can be challenging. *Model merging* offers a solution to these challenges by combining multiple pretrained models into one model, giving it the combined abilities of each individual model, without any additional training.
+
+PEFT provides two methods for model merging:
+
+* [TIES-Merging](https://hf.co/papers/2306.01708) - TrIm, Elect, and Merge (TIES) is a three-step method for merging models. First, redundant parameters are trimmed, then conflicting signs are resolved into an aggregated vector, and finally the parameters whose signs are the same as the aggregate sign are averaged. This method takes into account that some values (redundant and sign disagreement) can degrade performance in the merged model.
+* [DARE](https://hf.co/papers/2311.03099) - Drop And REscale is a method that can be used to prepare for model merging methods like TIES. It works by randomly dropping parameters according to a drop rate and rescaling the remaining parameters. This helps to reduce the number of redundant and potentially interfering parameters among multiple models.
+
+Models are merged with the [`~LoraModel.add_weighted_adapter`] method, and the specific model merging method is specified in the `combination_type` parameter. This guide will show you how to merge models with TIES, DARE, and a combination of both TIES and DARE.
+
+## TIES
+
```

Comment:
It might be worth adding a note of caution for TIES which might not be an issue here but for the full model, it can be an issue. I discussed this in my previous comment on the pull request which adds these merging features. 

"As far as I remember, there were no special tricks when using TIES. Typically, it is important to take care of the shared parameters when merging full models. Sometimes the embedding layers of different finetunes have a few additional tokens added and they can be different across different finetunes. Say a math model adds "=" and a coding model adds "{" as additional tokens that will occupy the V+1 index in vocabulary but so in such cases the embedding layers should be merged carefully because different token embedding can be merged together. I feel like this is something people typically miss but should not be an issue with PEFT methods. For merging Lora I have tried merging BA and then decomposing it back which seems to work fine. However, I have not tried merging A and B matrices separately but I feel that it should not be much worse. How big is the performance gap between these two ways of merging Lora's in your experience?"

Reply:
Awesome, added a note about this! 👍